lib/set.js

```diff
@@ -2,74 +2,140 @@
+        this._set.add(value);
+
+        this._hasRef |= isRef;
+
         return this;
     }
 
     merge(add, remove) {
 
-        for (let i = 0; i < add._set.length; ++i) {
-            this.add(add._set[i]);
+        for (const item of add._set) {
+            this.add(item);
         }
 
-        for (let i = 0; i < remove._set.length; ++i) {
-            this.remove(remove._set[i]);
+        for (const item of remove._set) {
+            this.remove(item);
         }
 
         return this;
     }
 
     remove(value) {
 
-        this._set = this._set.filter((item) => value !== item);
+        this._set.delete(value);
         return this;
     }
 
     has(value, state, options, insensitive) {
 
-        for (let i = 0; i < this._set.length; ++i) {
-            let items = this._set[i];
+        const hasValue = this._set.has(value);
+        if (hasValue) {
+            return hasValue;
+        }
 
-            if (state && Ref.isRef(items)) { // Only resolve references if there is a state, otherwise it's a merge
-                items = items(state.reference || state.parent, options);
+        const extendedCheck = internals.extendedCheckForValue(value, insensitive);
+        if (!extendedCheck) {
+            if (state && this._hasRef) {
+                for (let item of this._set) {
+                    if (Ref.isRef(item)) {
+                        item = item(state.reference || state.parent, options);
+                        if (value === item || (Array.isArray(item) && item.includes(value))) {
+                            return true;
+                        }
+                    }
+                }
             }
 
-            if (!Array.isArray(items)) {
-                items = [items];
+            return false;
+        }
+
+        return this._has(value, state, options, extendedCheck);
+    }
+
+    _has(value, state, options, check) {
+
+        const checkRef = !!(state && this._hasRef);
+
+        const isReallyEqual = function (item) {
+
+            if (value === item) {
+                return true;
             }
 
-            for (let j = 0; j < items.length; ++j) {
-                const item = items[j];
-                if (typeof value !== typeof item) {
-                    continue;
-                }
+            return check(item);
+        };
 
-                if (value === item ||
-                    (value instanceof Date && item instanceof Date && value.getTime() === item.getTime()) ||
-                    (insensitive && typeof value === 'string' && value.toLowerCase() === item.toLowerCase()) ||
-                    (Buffer.isBuffer(value) && Buffer.isBuffer(item) && value.length === item.length && value.toString('binary') === item.toString('binary'))) {
+        for (let item of this._set) {
+            if (checkRef && Ref.isRef(item)) { // Only resolve references if there is a state, otherwise it's a merge
+                item = item(state.reference || state.parent, options);
```

Comment:
The missing coverage is because `state.reference` is always falsy.

I'm not sure how to add a test that exercises this code. Any  ideas?